src/main/scala/com/cognite/sdk/scala/v1/dataModelProperties.scala

```diff
@@ -3,53 +3,86 @@
 
 package com.cognite.sdk.scala.v1
 
-import java.time.{LocalDate, ZonedDateTime}
-
-sealed trait PropertyType
-
-sealed trait PropertyTypePrimitive extends PropertyType
-final case class BooleanProperty(value: Boolean) extends PropertyTypePrimitive
-final case class Int32Property(value: Int) extends PropertyTypePrimitive
-final case class Int64Property(value: Long) extends PropertyTypePrimitive
-final case class Float32Property(value: Float) extends PropertyTypePrimitive
-final case class Float64Property(value: Double) extends PropertyTypePrimitive
-final case class StringProperty(value: String) extends PropertyTypePrimitive
-
-final case class ArrayProperty[+A <: PropertyTypePrimitive](values: Vector[A]) extends PropertyType
-
-final case class TimeStampProperty(value: ZonedDateTime) extends PropertyType
-final case class DateProperty(value: LocalDate) extends PropertyType
-
-// These types below are treated as string for now
-final case class DirectRelationProperty(value: String) extends PropertyType
-final case class GeometryProperty(value: String) extends PropertyType
-final case class GeographyProperty(value: String) extends PropertyType
-
-object PropertyName {
-  val Boolean = "boolean"
-  val Int32 = "int32"
-  val Int64 = "int64"
-  val Int = "int"
-  val Bigint = "bigint"
-  val Float32 = "float32"
-  val Float64 = "float64"
-  val Numeric = "numeric"
-  val Text = "text"
-  val Timestamp = "timestamp"
-  val Date = "date"
-
-  val ArrayText = "text[]"
-  val ArrayBoolean = "boolean[]"
-  val ArrayInt32 = "int32[]"
-  val ArrayInt64 = "int64[]"
-  val ArrayInt = "int[]"
-  val ArrayBigint = "bigint[]"
-  val ArrayFloat32 = "float32[]"
-  val ArrayFloat64 = "float64[]"
-  val ArrayNumeric = "numeric[]"
+sealed abstract class PropertyType {
+
+  @SuppressWarnings(Array("org.wartremover.warts.PlatformDefault"))
+  def code: String =
+    toString.replaceAll("(.)([A-Z])", "$1_$2").toLowerCase
+}
+
+object PropertyType {
+  val values: Seq[PropertyType] = Seq[PropertyType](
+    Boolean,
+    Int,
+    Bigint,
+    Float32,
+    Float64,
+    Numeric,
+    Text,
+    Json,
+    Timestamp,
+    Date,
+    DirectRelation,
+    Geometry,
+    Geography
+  ) ++
+    Array.values
+
+  private val _valuesMap: Map[String, PropertyType] =
+    values.map(t => t.code -> t).toMap
+
+  def fromCode(code: String): Option[PropertyType] =
+    _valuesMap.get(code)
+
+  case object Boolean extends PropertyType
+  case object Int extends PropertyType
+  case object Bigint extends PropertyType
+  case object Float32 extends PropertyType
+  case object Float64 extends PropertyType
+  case object Numeric extends PropertyType
+  case object Text extends PropertyType
+  case object Json extends PropertyType
+  case object Timestamp extends PropertyType
+  case object Date extends PropertyType
```

Comment:
Not sure I get the new implementation, if we switch from case class to object how can we retrieve the "value" with with corresponding type and use it in Spark Data Source ? 🤔

Reply:
it was a renaming... `PropertyType` is what `PropertyName` used to be (an enum with all the property types)

the old `PropertyType` (the one that held the value for encoding/decoding) will be reintroduced as `DataModelProperty` in the instances PR.